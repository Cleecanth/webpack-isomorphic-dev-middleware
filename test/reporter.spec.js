'use strict';

const fs = require('fs');
const pTry = require('p-try');
const express = require('express');
const request = require('supertest');
const webpackIsomorphicDevMiddleware = require('../');
const createCompiler = require('./util/createCompiler');
const createOutputStream = require('./util/createOutputStream');
const configClientBasic = require('./configs/client-basic');
const configServerBasic = require('./configs/server-basic');
const configServerWithHashes = require('./configs/server-with-hashes');

describe('output', () => {
    const originalStderr = process.stderr;

    afterEach(() => { process.stderr = originalStderr; });
    afterEach(() => createCompiler.teardown());

    it('should report stats only once by default', () => {
        const app = express();
        const compiler = createCompiler(configClientBasic, configServerBasic);
        const outputStream = createOutputStream();

        app.use(webpackIsomorphicDevMiddleware(compiler, {
            watchOptions: { report: { output: outputStream } },
        }));

        return pTry(() => (
            request(app)
            .get('/client.js')
            .expect(200)
        ))
        .then(() => new Promise((resolve) => {
            fs.writeFileSync(configClientBasic.entry, fs.readFileSync(configClientBasic.entry));
            compiler.once('begin', resolve);
        }))
        .then(() => (
            request(app)
            .get('/client.js')
            .expect(200)
        ))
        .then(() => {
            expect(outputStream.getReportOutput()).toMatchSnapshot();
        });
    });

    it('should not report anything if watchOptions.report is falsy', () => {
        const app = express();
        const compiler = createCompiler(configClientBasic, configServerWithHashes);
        const outputStream = createOutputStream();

        process.stderr = outputStream;

        app.use(webpackIsomorphicDevMiddleware(compiler, {
            watchOptions: { report: null },
        }));

        return request(app)
        .get('/client.js')
        .expect(200)
        .then(() => {
            expect(outputStream.getReportOutput()).toMatchSnapshot();
        });
    });

    describe('human errors', () => {
        it('should warn if hashes are being used in webpack config', () => {
            const app = express();
            const compiler = createCompiler(configClientBasic, configServerWithHashes);
            const outputStream = createOutputStream();

            app.use(webpackIsomorphicDevMiddleware(compiler, {
                watchOptions: { report: { output: outputStream } },
            }));

            return request(app)
            .get('/client.js')
            .expect(200)
            .then(() => {
                expect(outputStream.getReportOutput()).toMatchSnapshot();
            });
        });

        it('should not warn about hashes are being used in webpack config if options.memoryFs = false', () => {
            const app = express();
            const compiler = createCompiler(configClientBasic, configServerWithHashes);
            const outputStream = createOutputStream();

            app.use(webpackIsomorphicDevMiddleware(compiler, {
                memoryFs: false,
                watchOptions: { report: { output: outputStream } },
            }));

            return request(app)
            .get('/client.js')
            .expect(200)
            .then(() => {
                expect(outputStream.getReportOutput()).toMatchSnapshot();
            });
        });

        it('should not check human errors if options.watchOptions.report is falsy', () => {
            const app = express();
            const compiler = createCompiler(configClientBasic, configServerWithHashes);
            const outputStream = createOutputStream();

            process.stderr = outputStream;

            app.use(webpackIsomorphicDevMiddleware(compiler, {
                memoryFs: false,
                watchOptions: { report: null },
            }));

            return request(app)
            .get('/client.js')
            .expect(200)
            .then(() => {
                expect(outputStream.getReportOutput()).toMatchSnapshot();
            });
        });
    });
});
