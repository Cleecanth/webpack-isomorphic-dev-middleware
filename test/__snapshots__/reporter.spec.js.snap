// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`human errors should not check human errors if options.report is false 1`] = `
"● Compiling...
✔ Compilation succeeded (10ms)

     CLIENT                             
    ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    Asset Size Chunks Chunk Names
    client.js  x.xx kB       0  [emitted]  main

     SERVER                             
    ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    Asset Size Chunks Chunk Names
    output.a1aaaaaaaaaaaaaaaaaaaaaaa.js  x.xx kB       0  [emitted]  main

"
`;

exports[`human errors should not check human errors if options.report.humanErrors is false 1`] = `
"● Compiling...
✔ Compilation succeeded (10ms)

     CLIENT                             
    ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    Asset Size Chunks Chunk Names
    client.js  x.xx kB       0  [emitted]  main

     SERVER                             
    ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    Asset Size Chunks Chunk Names
    output.a1aaaaaaaaaaaaaaaaaaaaaaa.js  x.xx kB       0  [emitted]  main

"
`;

exports[`human errors should not warn about hashes are being used in webpack config if options.memoryFs is false 1`] = `
"● Compiling...
✔ Compilation succeeded (10ms)

     CLIENT                             
    ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    Asset Size Chunks Chunk Names
    client.js  x.xx kB       0  [emitted]  main

     SERVER                             
    ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    Asset Size Chunks Chunk Names
    output.a1aaaaaaaaaaaaaaaaaaaaaaa.js  x.xx kB       0  [emitted]  main

"
`;

exports[`human errors should warn if hashes are being used in webpack config 1`] = `
"● Compiling...
✔ Compilation succeeded (10ms)

     CLIENT                             
    ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    Asset Size Chunks Chunk Names
    client.js  x.xx kB       0  [emitted]  main

     SERVER                             
    ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    Asset Size Chunks Chunk Names
    output.a1aaaaaaaaaaaaaaaaaaaaaaa.js  x.xx kB       0  [emitted]  main

WARN: Assets with an hash in its name was detected on the server:
- output.a1aaaaaaaaaaaaaaaaaaaaaaa.js

This is known to cause memory leaks with webpack-dev-middleware's in-memory filesystem.
Either avoid having hashes in development or disable in-memory filesystem through the \`memoryFs\` option.
If you feel this was a false positive, please ignore this warning.

"
`;

exports[`should report stats only once by default 1`] = `
"● Compiling...
✔ Compilation succeeded (10ms)

     CLIENT                             
    ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    Asset Size Chunks Chunk Names
    client.js  x.xx kB       0  [emitted]  main

     SERVER                             
    ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    Asset Size Chunks Chunk Names
    server.js  x.xx kB       0  [emitted]  main

● Compiling...
✔ Compilation succeeded (10ms)
"
`;
